core/src/ledger_cleanup_service.rs

```diff
@@ -121,63 +136,64 @@ impl LedgerCleanupService {
-    ///   with `first_slot_to_purge`, the two Slot values represent the
-    ///   range of the clean up.
-    /// - `total_shreds` (u64): the total estimated number of shreds before the
-    ///   `root`.
+    /// A helper function to `cleanup_ledger` which returns a struct describing
+    /// the details of a potential cleanup to perform.
     fn find_slots_to_clean(
         blockstore: &Arc<Blockstore>,
         root: Slot,
+        previous_last_purge_slot: Slot,
         max_ledger_shreds: u64,
-    ) -> (bool, Slot, Slot, u64) {
+    ) -> CleanupDetails {
         let mut total_slots = Vec::new();
         let mut iterate_time = Measure::start("iterate_time");
-        let mut total_shreds = 0;
+        let mut total_num_shreds = 0;
         let mut first_slot = 0;
         for (i, (slot, meta)) in blockstore.slot_meta_iterator(0).unwrap().enumerate() {
             if i == 0 {
                 first_slot = slot;
                 debug!("purge: searching from slot: {}", slot);
             }
             // Not exact since non-full slots will have holes
-            total_shreds += meta.received;
+            total_num_shreds += meta.received;
             total_slots.push((slot, meta.received));
             if slot > root {
                 break;
             }
         }
         iterate_time.stop();
         info!(
-            "first_slot={} total_slots={} total_shreds={} max_ledger_shreds={}, {}",
+            "first_slot={} total_slots={} total_num_shreds={} max_ledger_shreds={}, {}",
             first_slot,
             total_slots.len(),
-            total_shreds,
+            total_num_shreds,
             max_ledger_shreds,
             iterate_time
         );
-        if (total_shreds as u64) < max_ledger_shreds {
-            return (false, 0, 0, total_shreds);
+
+        // The coding shreds for slots <= previous_last_purge_slot should have
+        // been cleaned by previous run(s) so start the coding shred cleanup
+        // at the slot immediately after where the previous cleanup ended
+        let mut cleanup_details = CleanupDetails {
+            coding_shred_cleanup_range: Some((previous_last_purge_slot + 1, root)),
+            total_num_shreds,
+            ..CleanupDetails::default()
+        };
+
+        if (total_num_shreds as u64) < max_ledger_shreds {
+            return cleanup_details;
         }
-        let mut num_shreds_to_clean = 0;
-        let mut lowest_cleanup_slot = total_slots[0].0;
+
+        let mut num_shreds_to_keep = 0;
+        let mut max_purge_slot = total_slots[0].0;
         for (slot, num_shreds) in total_slots.iter().rev() {
-            num_shreds_to_clean += *num_shreds as u64;
-            if num_shreds_to_clean > max_ledger_shreds {
-                lowest_cleanup_slot = *slot;
+            num_shreds_to_keep += *num_shreds as u64;
+            if num_shreds_to_keep > max_ledger_shreds {
+                max_purge_slot = *slot;
                 break;
```

Comment:
Re-reading this code, `num_shreds_to_clean` is kind of confusing.

This loop looks at all slots starting from the newest, and sums up shreds in slots until we cross the `max_ledger_shreds` threshold. So, we are actually counting how many shreds we'll be keeping. Additionally, the slot we stop at will be the last slot that is purged in the all-column purge